File: controlplane/internal/query/deactivation_enforcement_db_test.go
package query_test

import (
	"context"
	"strings"
	"testing"

	"github.com/ridi-oss/proxy-monster/controlplane/internal/dbtest"
	"github.com/ridi-oss/proxy-monster/controlplane/internal/identity"
	pb "github.com/ridi-oss/proxy-monster/controlplane/internal/pb"
)

// DeactivationEnforcementDbTest.kt — 113 LOC, 4 cases (06-query-decision.md §7, step 5).
//
// The authoritative fail-closed regression for deprovisioning under the enforcement engine.
//
// 🔒 WHY THE OBVIOUS SIGNAL IS THE WRONG ONE. "RoleResolver erased their roles" is FAIL-OPEN here, and
// that inversion is the whole point of the suite: decideQuery builds column actions only from
// role-attached policies, and a statement that produces NO masks is an ALLOW (step 26). So a
// deactivated principal whose role set collapses to the empty set gets *no* masked-column policy to
// match, *no* mask, and therefore a cleartext ALLOW on the very query that MASK'd while they were
// employed — MORE access after deprovision, not less. Step 5's structural DENY, placed before role
// resolution ever runs, is what closes that.
//
// Each case therefore drives the decide → execute → mask harness end to end, not
// `RoleResolver.resolve` in isolation: checking only that the role set went empty is precisely the
// signal that is fail-open.
func TestDeactivationEnforcementDb(t *testing.T) {
	fx := newEnforcementFixture(t, dbtest.EnginePostgres)
	ctx := context.Background()
	userGroups := identity.NewUserGroupStore(fx.Store.Pool)

	// grantAnalystRole gives `principal` the fixture's seeded `analyst` (MASK-on-rrn) role via a
	// DIRECT assignment — the Kotlin's private helper of the same name, over the production store.
	grantAnalystRole := func(t *testing.T, principal string) {
		t.Helper()
		assignExistingRole(t, fx, principal, dbtest.FixtureRole)
	}

	// deprovision is `fx.userGroupStore.createUser(AppUserInput(principal), …)` +
	// `setUserActive(principal, false)`: an app_user row now exists for this principal and is
	// deactivated — the SCIM active=false / IdP-liveness-failure path.
	//
	// ⚠️ TODO(A3): the Kotlin calls UserGroupStore.createUser/setUserActive, which internal/identity has
	// not ported (it has IsDeactivated and RolesForPrincipal only). The seeder writes the same two rows;
	// re-point at the production methods when they land — they own the revocation side effects too.
	deprovision := func(t *testing.T, principal string) {
		t.Helper()
		fx.Seed.User(principal)
		fx.Seed.SetUserActive(principal, false)
	}

	// `a deactivated principal is denied instead of falling open to cleartext`
	t.Run("a deactivated principal is denied instead of falling open to cleartext", func(t *testing.T) {
		const principal = "deactivated-analyst-1@example.com"
		grantAnalystRole(t, principal)

		// Sanity: while active, the same query MASKs the PII column as expected.
		active := runAs(t, fx, principal, "select id, rrn from users order by id")
		if action(active) != pb.EnfAction_MASK {
			t.Fatalf("sanity: query must MASK while the principal is active; got %v (%s)",
				action(active), respReason(active))
		}

		// Deprovision. Without the fail-closed gate, RoleResolver would resolve to the empty set, no
		// column policy would match, and the decision would ALLOW — cleartext rrn leaking to a
		// deprovisioned user.
		deprovision(t, principal)
		deactivated, err := userGroups.IsDeactivated(ctx, principal)
		if err != nil || !deactivated {
			t.Fatalf("isDeactivated(%s) = %v, %v — want true", principal, deactivated, err)
		}

		r := runAs(t, fx, principal, "select id, rrn from users order by id")
		if action(r) != pb.EnfAction_DENY {
			t.Fatalf("a deactivated principal must be denied, not fall open to ALLOW; got %v", action(r))
		}
		if len(r.Rows) != 0 {
			t.Fatalf("a DENY must not return rows (no cleartext rrn leak), got %d", len(r.Rows))
		}
		if !strings.Contains(strings.ToLower(respReason(r)), "deprovision") {
			t.Errorf("deny reason should call out deprovisioning: %s", respReason(r))
		}
	})

	// `a deactivated principal is denied even for a non-sensitive lineage query`
	t.Run("a deactivated principal is denied even for a non-sensitive lineage query", func(t *testing.T) {
		const principal = "deactivated-nonsensitive@example.com"
		deprovision(t, principal)

		// Even a query touching no PII at all (would otherwise ALLOW cleanly) must still be denied —
		// the deactivation gate sits before the lineage triad, not folded into it.
		r := runAs(t, fx, principal, "select id, region from users order by id")
		if action(r) != pb.EnfAction_DENY {
			t.Fatalf("deactivation must deny regardless of the query's own sensitivity; got %v", action(r))
		}
		if len(r.Rows) != 0 {
			t.Fatalf("a DENY must not return rows, got %d", len(r.Rows))
		}
	})

	// 🔒 `a deactivated principal is denied even a READONLY_META passthrough (the gate dominates passthrough)`
	//
	// INV-A6-8. This is the case 06-query-decision.md names as the pin for the ordering: step 5 sits
	// BEFORE step 14's metadata/session passthrough dispatch, so a deprovisioned principal cannot ride
	// a `readonly-meta` passthrough to an ALLOW. A regression that moved the gate after passthrough
	// classification would silently reopen it, and ONLY this case would catch it — the ALLOW half above
	// it proves the very same statement passes through for the very same principal while active, so the
	// DENY below cannot be attributed to anything else.
	t.Run("a deactivated principal is denied even a READONLY_META passthrough (the gate dominates passthrough)", func(t *testing.T) {
		const principal = "deactivated-passthrough@example.com"
		// The datasource.connect gate runs ahead of passthrough classification too, so `select
		// version()` needs a connect grant to ALLOW even while active — grant `analyst` (which has one)
		// so this isolates the deactivation gate, not the connect gate.
		grantAnalystRole(t, principal)

		// `select version()` is admitted as a READONLY_META passthrough (no lineage, no policy); for any
		// non-deactivated, connected principal it ALLOWs straight through. This principal has no
		// app_user row yet, so isDeactivated is false.
		activePass := runAs(t, fx, principal, "select version()")
		if action(activePass) != pb.EnfAction_ALLOW {
			t.Fatalf("sanity: a READONLY_META statement passes through while active; got %v (%s)",
				action(activePass), respReason(activePass))
		}

		deprovision(t, principal)

		r := runAs(t, fx, principal, "select version()")
		if action(r) != pb.EnfAction_DENY {
			t.Fatalf("INV-A6-8 VIOLATED: deactivation must dominate passthrough classification; got %v",
				action(r))
		}
		if len(r.Rows) != 0 {
			t.Fatalf("a DENY must not return rows, got %d", len(r.Rows))
		}

		// ⚠️ DELIBERATE STRENGTHENING over the Kotlin, and the ONLY assertion in this case that actually
		// bites. MEASURED by mutation: moving step 5 to sit after step 14's passthrough dispatch leaves
		// the two assertions above PASSING, because the production RoleResolver ALSO short-circuits a
		// deactivated principal to the empty role set (RoleResolver.kt:45-54) — so the reordered
		// decision still DENYs, at step 12's `datasource.connect` gate, with "no access to datasource".
		// A DENY for the wrong reason is not the invariant: this fixture's connect grant happens to be
		// role-scoped, and on a datasource whose connect is role-AGNOSTIC (a `permit(principal, action ==
		// datasource.connect, resource)` policy, or the shipped system:development preset) that
		// substitute deny does not exist and the passthrough ALLOWs.
		//
		// So the verdict alone cannot distinguish "the gate dominated the passthrough" from "the empty
		// role set happened to fail a later gate" — which is the exact fail-open signal this suite's own
		// doc comment warns against. Asserting the deny is the DEPROVISIONED one closes that.
		if !strings.Contains(strings.ToLower(respReason(r)), "deprovision") {
			t.Errorf("INV-A6-8 VIOLATED: the DENY must be the deprovisioning deny (step 5), not a "+
				"later gate denying because the role set collapsed; got %s", respReason(r))
		}
	})

	// `reactivating a principal restores normal enforcement`
	t.Run("reactivating a principal restores normal enforcement", func(t *testing.T) {
		const principal = "reactivated-analyst@example.com"
		grantAnalystRole(t, principal)
		deprovision(t, principal)
		if got := action(runAs(t, fx, principal, "select id, rrn from users order by id")); got != pb.EnfAction_DENY {
			t.Fatalf("decision while deactivated = %v, want DENY", got)
		}

		fx.Seed.SetUserActive(principal, true)
		r := runAs(t, fx, principal, "select id, rrn from users order by id")
		if action(r) != pb.EnfAction_MASK {
			t.Fatalf("reactivation must restore normal (masked) enforcement, not a stuck DENY; got %v (%s)",
				action(r), respReason(r))
		}
	})
}

File: control-plane/src/test/kotlin/com/ridi/oss/proxymonster/controlplane/DeactivationEnforcementDbTest.kt
package com.ridi.oss.proxymonster.controlplane

import com.ridi.oss.proxymonster.grpc.EnfAction
import com.ridi.oss.proxymonster.controlplane.support.EnforcementFixture
import com.ridi.oss.proxymonster.controlplane.support.requireDockerOrSkip
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The authoritative fail-closed regression for deprovisioning under the enforcement engine
 * (RoleResolver erasing roles is the wrong signal — `decideQuery` only builds
 * column actions from role-attached policies, and [PolicyEvaluator] ALLOWs when no masks result, so
 * a deactivated principal's role set collapsing to empty flips a MASK'd query to cleartext ALLOW —
 * MORE access after deprovision). [decideQuery] emits a structural DENY for a deactivated
 * principal before role resolution ever runs, closing that hole. This exercises [runEnforcedForTest]
 * end-to-end (decide -> broker -> mask), not just [RoleResolver.resolve] in isolation — checking only
 * the empty role set is exactly the signal that is fail-open.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class DeactivationEnforcementDbTest {
    private lateinit var fx: EnforcementFixture

    @BeforeAll
    fun setup() {
        requireDockerOrSkip()
        fx = EnforcementFixture.postgres()
    }

    /** Grant [principal] the fixture's seeded `analyst` (MASK-on-ssn) role via a direct assignment. */
    private fun grantAnalystRole(principal: String) {
        val roleId = fx.policyStore.listRoles().first { it.name == fx.role }.id
        fx.policyStore.createAssignment(RoleAssignmentInput(principal, roleId))
    }

    @Test
    fun `a deactivated principal is denied instead of falling open to cleartext`() {
        val principal = "deactivated-analyst-1@example.com"
        grantAnalystRole(principal)

        // Sanity: while active, the same query MASKs the PII column as expected.
        val active = fx.run("select id, ssn from users order by id", principal = principal)
        assertEquals(EnfAction.MASK, active.decision, "sanity: query must MASK while the principal is active")

        // Deprovision: an app_user row now exists for this principal and is deactivated (the SCIM
        // active=false / IdP-liveness-failure path). Without the fail-closed gate, RoleResolver would
        // resolve to the empty set, no column policy would match, and PolicyEvaluator would ALLOW —
        // cleartext ssn leaking to a deprovisioned user.
        fx.userGroupStore.createUser(AppUserInput(principal = principal), fx.tokenStore, fx.accessStore, fx.daemonSessionStore)
        fx.userGroupStore.setUserActive(principal, false)
        assertTrue(fx.userGroupStore.isDeactivated(principal))

        val r = fx.run("select id, ssn from users order by id", principal = principal)
        assertEquals(EnfAction.DENY, r.decision, "a deactivated principal must be denied, not fall open to ALLOW")
        assertTrue(r.rows.isEmpty(), "a DENY must not return rows (no cleartext ssn leak)")
        assertTrue(
            r.denyReason?.contains("deprovision", ignoreCase = true) == true,
            "deny reason should call out deprovisioning: ${r.denyReason}",
        )
    }

    @Test
    fun `a deactivated principal is denied even for a non-sensitive lineage query`() {
        val principal = "deactivated-nonsensitive@example.com"
        fx.userGroupStore.createUser(AppUserInput(principal = principal), fx.tokenStore, fx.accessStore, fx.daemonSessionStore)
        fx.userGroupStore.setUserActive(principal, false)

        // Even a query touching no PII at all (would otherwise ALLOW cleanly) must still be denied —
        // the deactivation gate sits before the lineage triad, not folded into it.
        val r = fx.run("select id, region from users order by id", principal = principal)
        assertEquals(EnfAction.DENY, r.decision, "deactivation must deny regardless of the query's own sensitivity")
        assertTrue(r.rows.isEmpty())
    }

    @Test
    fun `a deactivated principal is denied even a READONLY_META passthrough (the gate dominates passthrough)`() {
        val principal = "deactivated-passthrough@example.com"
        // The datasource.connect gate runs ahead of passthrough classification too, so a
        // `select version()` needs a connect grant to ALLOW even while active — grant `analyst` (which
        // has one) so the sanity check below isolates the deactivation gate, not the connect gate.
        grantAnalystRole(principal)
        // `select version()` is admitted as a READONLY_META passthrough (no lineage, no policy) — for
        // any non-deactivated, connected principal it ALLOWs straight through. Sanity-check that first
        // (this principal has no app_user row yet, so isDeactivated is false).
        val activePass = fx.run("select version()", principal = principal)
        assertEquals(EnfAction.ALLOW, activePass.decision, "sanity: a READONLY_META statement passes through while active")

        fx.userGroupStore.createUser(AppUserInput(principal = principal), fx.tokenStore, fx.accessStore, fx.daemonSessionStore)
        fx.userGroupStore.setUserActive(principal, false)

        // The SAME passthrough statement must now DENY: the deactivation gate is placed BEFORE
        // adm.passthroughClass, so it dominates READONLY_META/TX_CONTROL/SESSION_MUTATING passthroughs.
        // A regression moving the gate after passthrough classification would silently reopen this.
        val r = fx.run("select version()", principal = principal)
        assertEquals(EnfAction.DENY, r.decision, "deactivation must dominate passthrough classification")
        assertTrue(r.rows.isEmpty())
    }

    @Test
    fun `reactivating a principal restores normal enforcement`() {
        val principal = "reactivated-analyst@example.com"
        grantAnalystRole(principal)
        fx.userGroupStore.createUser(AppUserInput(principal = principal), fx.tokenStore, fx.accessStore, fx.daemonSessionStore)
        fx.userGroupStore.setUserActive(principal, false)
        assertEquals(EnfAction.DENY, fx.run("select id, ssn from users order by id", principal = principal).decision)

        fx.userGroupStore.setUserActive(principal, true)
        val r = fx.run("select id, ssn from users order by id", principal = principal)
        assertEquals(EnfAction.MASK, r.decision, "reactivation must restore normal (masked) enforcement, not a stuck DENY")
    }
}
